### PR TITLE
MM-20684 old changes along with new tests added

### PR DESCRIPTION
**Pull request by KrishnaSindhur**
#### Ticket Link
Fixes https://github.com/mattermost/mattermost-server/issues/14739

### CODE REVIEW

**Comment by hanzei** [comment]
@KrishnaSindhur Does this PR replace https://github.com/mattermost/mattermost-server/pull/14810?

**Comment by KrishnaSindhur** [comment]
> #14810

yes

**Comment by KrishnaSindhur** [comment]
> @KrishnaSindhur Does this PR replace #14810?

yes it is

**Review by levb — APPROVED** [review]

**Comment by jasonblais** [comment]
@iomodo kind reminder to help with reviews :)

File: app/command_autocomplete.go [hunk 1]
```diff
@@ -174,18 +174,39 @@ func (a *App) parseArgument(commandArgs *model.CommandArgs, arg *model.Autocompl
...
+		parsed = changedParsed
+		toBeParsed = changedToBeParsed
 	}
 
 	return false, parsed, toBeParsed, suggestions
 }
 
 func parseNamedArgument(arg *model.AutocompleteArg, parsed, toBeParsed string) (found bool, alreadyParsed string, yetToBeParsed string, suggestion model.AutocompleteSuggestion) {
 	in := strings.TrimPrefix(toBeParsed, " ")
+	if len(arg.Name) == 0 && arg.Type == model.AutocompleteArgTypeBool {
```
Comment (iomodo) [inline app/command_autocomplete.go:192]: `parseNamedArgument` will not be called if `arg.Name` is empty. What is the point this check?

File: app/command_autocomplete.go [hunk 2]
```diff
@@ -174,18 +174,39 @@ func (a *App) parseArgument(commandArgs *model.CommandArgs, arg *model.Autocompl
...
+		if len(parsed) > 0 || len(toBeParsed) > 0 {
+			return true, parsed + toBeParsed, "", model.AutocompleteSuggestion{Complete: parsed + toBeParsed + " " + "false" + " ", Suggestion: "false", Hint: "", Description: arg.HelpText}
+		}
+		return true, parsed + toBeParsed, "", model.AutocompleteSuggestion{Complete: "false" + " ", Suggestion: "false", Hint: "", Description: arg.HelpText}
+	}
+
 	namedArg := "--" + arg.Name
 	if in == "" { //The user has not started typing the argument.
+		if arg.Type == model.AutocompleteArgTypeBool {
+			return true, parsed + toBeParsed, "", model.AutocompleteSuggestion{Complete: parsed + toBeParsed + namedArg + " " + "true" + " ", Suggestion: "true", Hint: "", Description: arg.HelpText}
```
Comment (iomodo) [inline app/command_autocomplete.go:202]: Do we need to suggest "true" here? Isn't just an argument name enough?
Example: `/command --randomize` will mean `randomize` argument is set(to true), and `/command` will mean `randomize` argument is not set(therefore is false).

File: app/command_autocomplete.go [hunk 3]
```diff
@@ -174,18 +174,39 @@ func (a *App) parseArgument(commandArgs *model.CommandArgs, arg *model.Autocompl
...
+		}
 		return true, parsed + toBeParsed, "", model.AutocompleteSuggestion{Complete: parsed + toBeParsed + namedArg + " ", Suggestion: namedArg, Hint: "", Description: arg.HelpText}
 	}
 	if strings.HasPrefix(strings.ToLower(namedArg), strings.ToLower(in)) {
+		if arg.Type == model.AutocompleteArgTypeBool {
+			return true, parsed + toBeParsed, "", model.AutocompleteSuggestion{Complete: parsed + toBeParsed + namedArg[len(in):] + " " + "true" + " ", Suggestion: "true", Hint: "", Description: arg.HelpText}
```
Comment (iomodo) [inline app/command_autocomplete.go:208]: same here, when bool type argument is named we do not need `true` and `false`, just `--argname` is enough

File: app/command_autocomplete.go [hunk 4]
```diff
@@ -286,3 +307,22 @@ func parseListItems(items []model.AutocompleteListItem, parsed, toBeParsed strin
 	}
 	return true, parsed + toBeParsed, "", suggestions
 }
+
+func parseBoolArgument(arg *model.AutocompleteArg, parsed, toBeParsed string) (found bool, alreadyParsed string, yetToBeParsed string, suggestion model.AutocompleteSuggestion) {
+	in := strings.TrimPrefix(toBeParsed, " ")
+	a := arg.Data.(*model.AutocompleteBoolArg)
+	if in == "" { //The user has not started typing the argument.
+		return true, parsed + toBeParsed, "", model.AutocompleteSuggestion{Complete: parsed + toBeParsed, Suggestion: "", Hint: a.Hint, Description: arg.HelpText}
```
Comment (iomodo) [inline app/command_autocomplete.go:315]: Shouldn't we have two suggestions here? `true` and `false`?

**Review by iomodo — CHANGES_REQUESTED** [review]
Looks good @KrishnaSindhur!
Couple of comments below

**Comment by isacikgoz** [comment]
Hey @KrishnaSindhur Are you still planning to work on this issue? Please let us know if there is something we can help with.

**Comment by KrishnaSindhur** [comment]
> Hey @KrishnaSindhur Are you still planning to work on this issue? Please let us know if there is something we can help with.

I am occupied with some other work, it will take some time to look into it

**Comment by isacikgoz** [comment]
Hey @KrishnaSindhur did you find some time to work on this ticket? Please let us know if you need some help.

**Comment by isacikgoz** [comment]
Hi @KrishnaSindhur, big thanks for your contribution! There hasn't been recent activity on this PR so I've marked it as inactive. Following our [inactive contribution process](https://developers.mattermost.com/contribute/getting-started/inactive-contributions/), the PR is now closed due to 30 days or more of inactivity.

However, if you're still interested working on the changes, let us know! Note that we [merged multiple repositories into one monorepo](https://github.com/mattermost/mattermost-server/issues/22420), which may require you to resubmit the changes. 

If you have any questions, don't hesitate to let us know, we're happy to help.

**Pull request closed without merging**